plutus-core/plutus-ir/src/PlutusIR/TypeCheck/Internal.hs

```diff
@@ -218,6 +221,58 @@ inferTypeM (Unwrap ann term)        = do
 inferTypeM (Error ann ty)           = do
     checkKindM ann ty $ Type ()
     normalizeTypeM $ void ty
+
+-- resTy ~> vResTy     vResTy = sop s_0 ... s_i ... s_n     s_i = [p_0 ... p_m]
+-- [check| G !- t_j : p_j]
+-- ----------------------------------------------------------------------------
+-- [infer| G !- constr resTy i t_0 ... t_n : vResTy]
+inferTypeM t@(Constr ann resTy i args) = do
```

Comment:
Is there really no way we could abstract this and the other one into a reusable function? Copy-pasting so much code feels very unsettling.